src/agent.rs

```diff
@@ -0,0 +1,483 @@
+use dpdk_sys::*;
+use std::collections::{BTreeMap, BTreeSet};
+use std::mem;
+use std::ptr::{self, NonNull};
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc, Mutex,
+};
+use tokio::{
+    runtime::{Builder, Runtime},
+    sync::mpsc,
+    task::{self, JoinHandle},
+};
+
+use crate::mbuf::Mbuf;
+use crate::proto::{ETHER_HDR_LEN, IP_NEXT_PROTO_UDP};
+use crate::udp::handle_ipv4_udp;
+use crate::{Error, Result};
+
+/// Burst size for `rte_tx_burst` and `rte_rx_burst`.
+const MAX_PKT_BURST: u16 = 32;
+/// Channel size for `TxAgent`.
+const TX_CHAN_SIZE: usize = 256;
+/// The capacity of a `TxBuffer`.
+const TX_BUF_SIZE: usize = 1024;
+/// The number of mbufs to flush a `TxBuffer`.
+#[allow(dead_code)]
+const TX_BUF_THRESH: usize = 8;
+
+/// Number of buckets in the hash table.
+const IP_FRAG_TABLE_BUCKET_NUM: u32 = 128;
+/// Number of entries per bucket (e.g. hash associativity). Should be power of two.
+const IP_FRAG_TABLE_BUCKET_SIZE: u32 = 16;
+/// Maximum number of entries that could be stored in the table. The value should be less
+/// or equal then `bucket_num` * `bucket_entries`.
+const IP_FRAG_TABLE_MAX_ENTRIES: u32 = 2048;
+
+/// An agent thread doing receiving.
+pub(crate) struct RxAgent {
+    /// A bool indicating whether the thread is running.
+    running: AtomicBool,
+    /// A set of (`port_id`, `queue_id`) to be polled.
+    tasks: Mutex<BTreeSet<(u16, u16)>>,
+}
+
+/// An agent thread doing sending.
+pub(crate) struct TxAgent {
+    /// Single-threaded Runtime.
+    rt: Option<Runtime>,
+    /// (`port_id`, `queue_id`) -> Task
+    tasks: Mutex<BTreeMap<(u16, u16), JoinHandle<()>>>,
+}
+
+/// Table holding fragmented packets.
+struct IpFragmentTable {
+    /// `rte_ip_frag_tbl` pointer.
+    tbl: NonNull<rte_ip_frag_tbl>,
+}
+
+/// Table holding packets to be deallocated.
+struct IpFragDeathRow {
+    /// `rte_ip_frag_death_row` pointer.
+    dr: NonNull<rte_ip_frag_death_row>,
+}
+
+#[allow(unsafe_code)]
+#[allow(clippy::missing_docs_in_private_items)]
+impl IpFragmentTable {
+    fn new(socket_id: u32) -> Result<Self> {
+        const MS_PER_S: u64 = 1000;
+        let max_cycles = unsafe { (rte_get_tsc_hz() + MS_PER_S - 1) / MS_PER_S * MS_PER_S };
+        let ptr = unsafe {
+            rte_ip_frag_table_create(
+                IP_FRAG_TABLE_BUCKET_NUM,
+                IP_FRAG_TABLE_BUCKET_SIZE,
+                IP_FRAG_TABLE_MAX_ENTRIES,
+                max_cycles,
+                socket_id as _,
+            )
+        };
+        let tbl = NonNull::new(ptr).ok_or(Error::NoMem)?;
+        Ok(Self { tbl })
+    }
+    fn as_mut_ptr(&mut self) -> *mut rte_ip_frag_tbl {
+        self.tbl.as_ptr()
+    }
+}
+
+#[allow(unsafe_code)]
+impl Drop for IpFragmentTable {
+    fn drop(&mut self) {
+        unsafe { rte_ip_frag_table_destroy(self.as_mut_ptr()) };
+    }
+}
+
+#[allow(unsafe_code)]
+#[allow(clippy::missing_docs_in_private_items)]
+impl IpFragDeathRow {
```

Comment:
Do we need a drop function to release the resource?

Reply:
Oops I forgot that.